### DISCUSSION TITLE
Displaying large (~2300x1700) images at 10Hz with minimal lag—is this possible with pyqtgraph?

### DISCUSSION THREAD

**Discussion by st-walker** [pyqtgraph/pyqtgraph#2798] [Q&A]
Firstly, I realise that what I am trying to do is may not be possible.

I want to use pyqtgraph to make a GUI for our free electron laser, in particular I want to look at the electron beam on a screen inserted in front of the electron beam.  The beam appears on the screen at a rep rate of 10Hz.  So we have new images every 0.1s, and the image is fairly large (about 2300x1700 pixels) greyscale (just a 2d numpy array of integers).  I would ideally like to show the image at 10Hz.  I tried this:

```python
widget = self.ui.screen_display_widget # this is a GraphicsLayoutWidget instance
image_plot = widget.addPlot()
image_plot.clear()
image_plot.hideAxis("left")
image_plot.hideAxis("bottom")
image = pg.ImageItem(autoDownsample=True, border="k")
```

I tried updating the image using a QTimer instance every 100ms, but this was really slow, so I decided to fire up a worker thread to make retrieving the image from the machine parallel to the main UI thread operation:

```python
class ScreenWatcher(QObject):
    image_signal = pyqtSignal(object)
    def __init__(self, machine):
        super().__init__()
        self.machine = machine
        self.screen_name = "OTRC.55.I1"

    def get_image(self):
        image = self.machine.sservice.get_image(self.screen_name)
        LOG.info("Reading image from: %s", self.screen_name)
        return image # .astype(np.float32)

    def run(self):
        while True:
            time.sleep(0.1)
            self.image_signal.emit(self.get_image())

    def update_screen_name(self, screen_name):
        self.screen_name = screen_name


```

Threading boiler plate:

```python
        screen_worker = ScreenWatcher(self.machine)
        screen_thread = QThread()
        screen_worker.moveToThread(screen_thread)
        screen_thread.started.connect(screen_worker.run)
        screen_worker.image_signal.connect(self.post_beam_image)
        screen_thread.start()
```

I then connect this signal `image_signal` to the following function in the main thread (the UI thread I guess since the UI is not thread safe):

```python
    def post_beam_image(self, image):
        items = self.image_plot.items
        assert len(items) == 1
        image_item = items[0]
        LOG.debug("Posting beam image...")
        image_item.setImage(image)
```

This is still really slow.   Resizing the window is really, really slow, and I would like it to be fast and responsive to the user.  So I can now only assume that it's the drawing of the image itself that is slow.  I tried to follow the advice in the `ImageItem` docs: useNumba, row-major, float32, no lookup tables, but it is still slow.  I didn't touch the levels stuff because to be honest I don't really understand it.  It is still slow.

The image on the screen really does not need to be perfect, it is just for the user to look at and see what is going on with the electron beam.   Almost of the image is usually blank, with the electron beam occupying a small region of the screen.

My questions:

1. Is what I am doing actually possible?  Can I display such large images at 10Hz responsively and without slowdown or is this just not realistic?  If it's possible how do I do it?
2. When I am resizing the main GUI window can I temporarily disable the redrawing of the image so as to make this faster and more user friendly?  I guess there must be some sort of signal emitted when the window is being resized?
3. Can I use a priori knowledge about the images to speed the whole thing up?  The image *mostly* doesn't change, for example, and it's mostly empty, the beam is small compared to the screen it is appearing on.
4. Can I preprocess the beam image in some way on the worker thread to speed this up further?

Thank you for your time and reading this post.

**Comment by pijyoi** [comment]
You didn't specify your data type. Would it be uint8, uint16? 
What is the output of `print(image.flags)`?
What version of pyqtgraph, numpy are you using?

On my ordinary laptop, without using numba nor downsampling:
`python .\pyqtgraph\examples\VideoSpeedTest.py --dtype uint8 --size=2300x1700 --lut` --> 103 fps
`python .\pyqtgraph\examples\VideoSpeedTest.py --dtype uint16 --size=2300x1700 --levels=0,10000 --lut` --> 44 fps
`python .\pyqtgraph\examples\VideoSpeedTest.py --dtype float --size=2300x1700 --levels=0,2 --lut` --> 31 fps

```python
    def run(self):
        while True:
            time.sleep(0.1)
            self.image_signal.emit(self.get_image())
```
Why the sleep for 100ms? Did you time how long `self.get_image()` takes?

>Is what I am doing actually possible? Can I display such large images at 10Hz responsively and without slowdown or is this just not realistic? 

Well, the `VideoSpeedTest` shows that even using float32 runs at 31fps, so I would say yes.

>If it's possible how do I do it?

If you could provide a minimal runnable example demonstrating/simulating what you are experience, then we could help you tweak it.
